include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+        virtual uint32_t read(IReadFile* file, void* buffer, size_t offset, size_t size) = 0;
+        virtual uint32_t write(IWriteFile* file, const void* buffer, size_t offset, size_t size) = 0;
+    };
+
+protected:
+    struct SRequestType;
+    class CAsyncQueue;
+
+    class SFutureBase
+    {
+        friend ISystem;
+
+    protected:
+        bool valid_flag = false;
+        SRequestType* request;
+        CAsyncQueue* asyncQ;
+
+    public:
+        // SFuture is non-copyable and non-movable
+        SFutureBase(const SFutureBase&) = delete;
+
+        ~SFutureBase()
+        {
+            request->set_skip();
+        }
+
+        bool ready() const { return !request || request->ready; }
+        bool valid() const { return valid_flag; }
+
+        void wait()
+        {
+            asyncQ->waitForRequestCompletion(request[0]);
+        }
+    };
+
+    template <typename T>
+    class SFuture : public SFutureBase
+    {
+        alignas(T) uint8_t storage[sizeof(T)];
+
+    public:
+        ~SFuture()
+        {
+            if (valid_flag)
+                getStorage()->~T();
+        }
+
+        T* getStorage() { return reinterpret_cast<T*>(storage); }
+        T get()
+        {
+            wait();
+            assert(valid_flag);
+            valid_flag = false;
+            T* ptr = getStorage();
+            return std::move(ptr[0]);
+        }
+    };
+
+    // @sadiuk add more request types if needed
+    enum E_REQUEST_TYPE
+    {
+        ERT_CREATE_READ,
+        ERT_CREATE_WRITE,
+        ERT_READ,
+        ERT_WRITE
+    };
+    template <E_REQUEST_TYPE RT>
+    struct SRequestParamsBase
+    {
+        static inline constexpr E_REQUEST_TYPE type = RT;
+        SFutureBase* future;
+        void* retval_storage;
+    };
+    struct SRequestParams_CREATE_READ : SRequestParamsBase<ERT_CREATE_READ>
+    {
+        std::filesystem::path filename;
+    };
+    struct SRequestParams_CREATE_WRITE : SRequestParamsBase<ERT_CREATE_WRITE>
+    {
+        std::filesystem::path filename;
+    };
+    struct SRequestParams_READ : SRequestParamsBase<ERT_READ>
+    {
+        IReadFile* file;
+        void* buffer;
+        size_t offset;
+        size_t size;
+    };
+    struct SRequestParams_WRITE : SRequestParamsBase<ERT_WRITE>
+    {
+        IWriteFile* file;
+        const void* buffer;
+        size_t offset;
+        size_t size;
+    };
+    struct SRequestType : SSkippableRequestBase
+    {
+        SFutureBase* future = nullptr;
+        void* retval_storage = nullptr;
+        E_REQUEST_TYPE type;
+        std::variant<
+            SRequestParams_CREATE_READ,
+            SRequestParams_CREATE_WRITE,
+            SRequestParams_READ,
+            SRequestParams_WRITE
+        > params;
+
+        void set_skip()
+        {
+            write_lock_guard<> lk(rwlock);
+            set_skip_no_lock();
+            future->request = nullptr;
+        }
```

Comment:
all of the logic should be in `SSkippableRequestBase` (and skip should be replaced with cancel or abort)